From 9463783088a354429704e8147076a8b3f9d42b24 Mon Sep 17 00:00:00 2001
From: Aradhya Bhatia <a-bhatia1@ti.com>
Date: Wed, 11 Jan 2023 22:39:30 +0530
Subject: [PATCH] mach-k3: am62a7: Add support to enable DSS QoS

Enable the Display Subsystem (DSS) QoS and make the orderID 8 such that
the DSS traffic is catered from the real-time queue of the DDR.

Signed-off-by: Aradhya Bhatia <a-bhatia1@ti.com>
---
 arch/arm/mach-k3/am62a7_init.c            | 16 ++++++++++++++++
 arch/arm/mach-k3/include/mach/am62a_qos.h | 20 ++++++++++++++++++++
 arch/arm/mach-k3/include/mach/hardware.h  |  1 +
 3 files changed, 37 insertions(+)
 create mode 100644 arch/arm/mach-k3/include/mach/am62a_qos.h

diff --git a/arch/arm/mach-k3/am62a7_init.c b/arch/arm/mach-k3/am62a7_init.c
index 75796ed166bf..c4364529c5cd 100644
--- a/arch/arm/mach-k3/am62a7_init.c
+++ b/arch/arm/mach-k3/am62a7_init.c
@@ -76,6 +76,19 @@ static void ctrl_mmr_unlock(void)
 	mmr_unlock(PADCFG_MMR1_BASE, 1);
 }
 
+static void setup_dss_qos(void)
+{
+	unsigned int channel;
+
+	/* DMA master port */
+	writel(0x76543210, (uintptr_t)QOS_DSS0_DMA_CBASS_GRP_MAP1);
+	writel(0xFEDCBA98, (uintptr_t)QOS_DSS0_DMA_CBASS_GRP_MAP2);
+
+	for (channel = 0; channel < QOS_DSS0_DMA_NUM_I_CH; channel++)
+		writel((QOS_DSS0_DMA_ORDER_ID << 4),
+		       (uintptr_t)QOS_DSS0_DMA_CBASS_MAP(channel));
+}
+
 void board_init_f(ulong dummy)
 {
 	struct udevice *dev;
@@ -166,6 +179,9 @@ void board_init_f(ulong dummy)
 		panic("DRAM init failed: %d\n", ret);
 #endif
 
+	/* DSS QoS setup */
+	setup_dss_qos();
+
 	printf("am62a_init: %s done\n", __func__);
 }
 
diff --git a/arch/arm/mach-k3/include/mach/am62a_qos.h b/arch/arm/mach-k3/include/mach/am62a_qos.h
new file mode 100644
index 000000000000..d9a0e305fc70
--- /dev/null
+++ b/arch/arm/mach-k3/include/mach/am62a_qos.h
@@ -0,0 +1,20 @@
+/* SPDX-License-Identifier: GPL-2.0+ */
+/*
+ * K3: AM62A QoS params definitions
+ *
+ * (C) Copyright (C) 2022-2023 Texas Instruments Incorporated - http://www.ti.com/
+ */
+#ifndef __ASM_ARCH_AM62A_QOS_PARAMS_H
+#define __ASM_ARCH_AM62A_QOS_PARAMS_H
+
+#define QOS_DSS0_DMA_ORDER_ID                   (8U)
+
+/* DSS QoS Registers */
+
+#define QOS_DSS0_DMA_BASE                       0x45D25000
+#define QOS_DSS0_DMA_NUM_I_CH                   4
+#define QOS_DSS0_DMA_CBASS_GRP_MAP1             (QOS_DSS0_DMA_BASE + 0x00)
+#define QOS_DSS0_DMA_CBASS_GRP_MAP2             (QOS_DSS0_DMA_BASE + 0x04)
+#define QOS_DSS0_DMA_CBASS_MAP(i)               (QOS_DSS0_DMA_BASE + 0x100 + (i) * 4)
+
+#endif  /* __ASM_ARCH_AM62A_QOS_PARAMS_H */
diff --git a/arch/arm/mach-k3/include/mach/hardware.h b/arch/arm/mach-k3/include/mach/hardware.h
index 6be69aa8bcfe..f52f79c474de 100644
--- a/arch/arm/mach-k3/include/mach/hardware.h
+++ b/arch/arm/mach-k3/include/mach/hardware.h
@@ -28,6 +28,7 @@
 
 #ifdef CONFIG_SOC_K3_AM62A7
 #include "am62a_hardware.h"
+#include "am62a_qos.h"
 #endif
 
 #ifdef CONFIG_SOC_K3_J784S4
-- 
2.39.0

